### ISSUE TITLE
test regression with latest SocketIO 2.1.2 using flask-socketio 3.1.1

### ISSUE THREAD

**Issue by anthraxx** [miguelgrinberg/Flask-SocketIO#858]
Im using the latest versions of all dependencies and encounter following issue while packaging for Arch Linux:

```python
============================= test session starts ==============================
platform linux -- Python 3.7.1, pytest-4.0.2, py-1.7.0, pluggy-0.8.0
rootdir: /build/python-flask-socketio/src/Flask-SocketIO-3.1.1, inifile:
collected 33 items

test_socketio.py .....F.FF........................                       [100%]

=================================== FAILURES ===================================
________ TestSocketIO.test_connect_class_based_query_string_and_headers ________

self = <test_socketio.TestSocketIO testMethod=test_connect_class_based_query_string_and_headers>

    def test_connect_class_based_query_string_and_headers(self):
        client = socketio.test_client(
            app, namespace='/ns', query_string='foo=bar',
            headers={'Authorization': 'Basic foobar'})
        received = client.get_received('/ns')
        self.assertEqual(len(received), 3)
        self.assertEqual(received[0]['args'], 'connected-ns')
>       self.assertEqual(received[1]['args'], '{"foo": ["bar"]}')
E       AssertionError: '{"foo": "bar"}' != '{"foo": ["bar"]}'
E       - {"foo": "bar"}
E       + {"foo": ["bar"]}
E       ?         +     +

test_socketio.py:571: AssertionError
_________ TestSocketIO.test_connect_namespace_query_string_and_headers _________

self = <test_socketio.TestSocketIO testMethod=test_connect_namespace_query_string_and_headers>

    def test_connect_namespace_query_string_and_headers(self):
        client = socketio.test_client(
            app, namespace='/test', query_string='foo=bar',
            headers={'My-Custom-Header': 'Value'})
        received = client.get_received('/test')
        self.assertEqual(len(received), 3)
        self.assertEqual(received[0]['args'], 'connected-test')
>       self.assertEqual(received[1]['args'], '{"foo": ["bar"]}')
E       AssertionError: '{"foo": "bar"}' != '{"foo": ["bar"]}'
E       - {"foo": "bar"}
E       + {"foo": ["bar"]}
E       ?         +     +

test_socketio.py:286: AssertionError
______________ TestSocketIO.test_connect_query_string_and_headers ______________

self = <test_socketio.TestSocketIO testMethod=test_connect_query_string_and_headers>

    def test_connect_query_string_and_headers(self):
        client = socketio.test_client(
            app, query_string='?foo=bar&foo=baz',
            headers={'Authorization': 'Bearer foobar'})
        received = client.get_received()
        self.assertEqual(len(received), 3)
        self.assertEqual(received[0]['args'], 'connected')
>       self.assertEqual(received[1]['args'], '{"foo": ["bar", "baz"]}')
E       AssertionError: '{"foo": "bar"}' != '{"foo": ["bar", "baz"]}'
E       - {"foo": "bar"}
E       + {"foo": ["bar", "baz"]}
E       ?         +     ++++++++

test_socketio.py:265: AssertionError
=============================== warnings summary ===============================
/usr/lib/python3.7/site-packages/werkzeug/datastructures.py:16
  /usr/lib/python3.7/site-packages/werkzeug/datastructures.py:16: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    from collections import Container, Iterable, MutableSet
  /usr/lib/python3.7/site-packages/werkzeug/datastructures.py:16: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    from collections import Container, Iterable, MutableSet
  /usr/lib/python3.7/site-packages/werkzeug/datastructures.py:16: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    from collections import Container, Iterable, MutableSet

/usr/lib/python3.7/site-packages/jinja2/utils.py:485
  /usr/lib/python3.7/site-packages/jinja2/utils.py:485: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    from collections import MutableMapping

/usr/lib/python3.7/site-packages/jinja2/runtime.py:318
  /usr/lib/python3.7/site-packages/jinja2/runtime.py:318: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    from collections import Mapping

/usr/lib/python3.7/site-packages/dns/hash.py:25
  /usr/lib/python3.7/site-packages/dns/hash.py:25: DeprecationWarning: dns.hash module will be removed in future versions. Please use hashlib instead.
    DeprecationWarning)

-- Docs: https://docs.pytest.org/en/latest/warnings.html
=============== 3 failed, 30 passed, 6 warnings in 0.82 seconds ================

```

**Comment by miguelgrinberg** [comment]
Oh, this is quite interesting, there is a difference in behavior between Python 3.7 and older versions.

```python
Python 3.6.6 (default, Jul 18 2018, 16:28:42)
[GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from werkzeug import ImmutableMultiDict
>>> dict(ImmutableMultiDict([('foo', 'bar'), ('foo', 'baz')]))
{'foo': ['bar', 'baz']}
```

```python
Python 3.7.1 (default, Nov 17 2018, 18:50:19)
[Clang 10.0.0 (clang-1000.10.44.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from werkzeug import ImmutableMultiDict
>>> dict(ImmutableMultiDict([('foo', 'bar'), ('foo', 'baz')]))
{'foo': 'bar'}
```

So this isn't something I can address from my side, it would seem as if Werkzeug had a regression. I'll look into it.

**Comment by miguelgrinberg** [comment]
Fixed in 3.1.2. Thanks for reporting the problem!
